bot/exts/info/doc/_batch_parser.py

```diff
@@ -0,0 +1,206 @@
+            )
+            await self._dev_log.send(embed=embed)
+
+
+class QueueItem(NamedTuple):
+    """Contains a doc_item and the BeautifulSoup object needed to parse it."""
+
+    doc_item: _cog.DocItem
+    soup: BeautifulSoup
+
+    def __eq__(self, other: Union[QueueItem, _cog.DocItem]):
+        if isinstance(other, _cog.DocItem):
+            return self.doc_item == other
+        return NamedTuple.__eq__(self, other)
+
+
+class ParseResultFuture(asyncio.Future):
+    """
+    Future with metadata for the parser class.
+
+    `user_requested` is set by the parser when a Future is requested by an user and moved to the front,
+    allowing the futures to only be waited for when clearing if they were user requested.
+
+    `result_set_time` provides the time at which the future's result has been set,
+    or -inf if the result hasn't been set yet
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.user_requested = False
+        self.result_set_time = float("inf")
+
+    def set_result(self, result: str, /) -> None:
+        """Set `self.result_set_time` to current time when the result is set."""
+        self.result_set_time = time.time()
+        super().set_result(result)
+
+
+class BatchParser:
+    """
+    Get the Markdown of all symbols on a page and send them to redis when a symbol is requested.
+
+    DocItems are added through the `add_item` method which adds them to the `_page_doc_items` dict.
+    `get_markdown` is used to fetch the Markdown; when this is used for the first time on a page,
+    all of the symbols are queued to be parsed to avoid multiple web requests to the same page.
+    """
+
+    def __init__(self):
+        self._queue: List[QueueItem] = []
+        self._page_doc_items: Dict[str, List[_cog.DocItem]] = defaultdict(list)
+        self._item_futures: Dict[_cog.DocItem, ParseResultFuture] = {}
+        self._parse_task = None
+
+        self.cleanup_futures_task = bot.instance.loop.create_task(self._clean_up_futures())
+
+        self.stale_inventory_notifier = StaleInventoryNotifier()
+
+    async def get_markdown(self, doc_item: _cog.DocItem) -> str:
+        """
+        Get the result Markdown of `doc_item`.
+
+        If no symbols were fetched from `doc_item`s page before,
+        the HTML has to be fetched and then all items from the page are put into the parse queue.
+
+        Not safe to run while `self.clear` is running.
+        """
+        if doc_item not in self._item_futures:
+            self._item_futures.update((item, ParseResultFuture()) for item in self._page_doc_items[doc_item.url])
+            self._item_futures[doc_item].user_requested = True
+
+            async with bot.instance.http_session.get(doc_item.url) as response:
+                soup = await bot.instance.loop.run_in_executor(
+                    None,
+                    BeautifulSoup,
+                    await response.text(encoding="utf8"),
+                    "lxml",
+                )
```

Comment:
If we get through the request I don't think this can really fail unless the data gets corrupted in some way. The HTML is auto generated and the parser is fairly generous in what it allows

Reply:
I think it wouldn't hurt to take the precaution, but it's not strictly necessary. I'll leave it up to you to decide.

Reply:
The code on master doesn't do any handling around it and it doesn't look like there were any issues with it. I'll leave it for now, with my points above in mind, and if an issue pops up in the future it can be addressed then

Reply:
Actually, this will fall under the broad exception handling added in 51a11cc4